### DISCUSSION TITLE
Issue with Zitadel Terraform Provider

### DISCUSSION THREAD

**Discussion by rgilmetdinov-i3** [zitadel/zitadel#8866] [Q&A]
Hello! Currently, there’s an issue with authenticating the Zitadel provider to the Zitadel instance.

Here’s the configuration:
```
main.tf
terraform {
  required_providers {
    zitadel = {
      source  = "zitadel/zitadel"
      version = "2.0.1"
    }
  }
}

provider "zitadel" {
  domain           = "mydomain.com"
  insecure         = "false"
  port             = "443"
  jwt_profile_json = <<EOF
    {
      "client_id": "service user with IAM Owner role",
      "client_secret": "generated secret of this user"
    }
EOF
}

resource "zitadel_org" "default" {
  name = "my-test-org"
}

data "zitadel_project" "default" {
  org_id     = data.zitadel_org.default.id
  project_id = "123456789012345678"
}

data "zitadel_org" "default" {
  id = "123456789012345678"
}

output "zitadel_org" {
  value = data.zitadel_org.default
}
```
When running terraform plan, I encounter the following error:
```
│ Error: failed to start zitadel client: PEM decode failed
│ 
│   with data.zitadel_org.default,
│   on main.tf line 10, in data "zitadel_org" "default":
│   10: data "zitadel_org" "default" {
```
Has anyone encountered this issue and found a solution? It’s possible that my jwt_profile_json block is incorrectly defined. And should I define the real ID of org?

Any advice would be greatly appreciated!

**Comment by sm3142** [comment]
While I personally have never used `jwt_profile_json`, only `jwt_profile_file`, the content presumably needs to be exactly the same. If you create a key for your Terraform service user and download it, it will contain something like this:
```
{
  "type": "serviceaccount",
  "keyId": "123456789012345678",
  "key": <PEM private key>,
  "expirationDate": "9999-12-31T23:59:59Z",
  "userId": "987654321987654321",
}
```
Have you tried pasting the verbatim contents of your key file between the EOF heredoc tags?

**Reply by rgilmetdinov-i3** [reply]
Hey @sm3142 thank you for your response!

I’ve tried what you suggested, and it seems I’ve made a bit of progress with that, but this time I got another error. I think it’s related to the org ID that I stored. Could that be the issue?
```
│ Error: error while getting org by id 292609476421952158: rpc error: code = Unknown desc = unexpected HTTP status code received from server: 464 (); malformed header: missing HTTP content-type
│ 
│   with data.zitadel_org.default,
│   on zitadel.tf line 10, in data "zitadel_org" "default":
│   10: data "zitadel_org" "default" {
```

**Reply by sm3142** [reply]
Haven't seen this error before to be honest. If I understand your intent hre correctly though, you want to create an organization and a project in that organization. In that case, the resource creation part would look more like this:

```
resource "zitadel_org" "my-test-org" {
  name = "my-test-org"
}

resource "zitadel_project" "my-test-project" {
  org_id     = zitadel_org.my-test-org.id
  name = "my-test-project"
}

output "zitadel_org" {
  value = zitadel_org.my-test-org
}

output "zitadel_project" {
  value = zitadel_project.my-test-project
}
```

(don't kill me for typos, I just jotted this down without testing. Also I just noticed that I haven't made the leap to the 2.x provider yet, so YMMV).

PS: In a simple scenario like this, one usually doesn't require data sources. A typical scenario for using a data source would be to get information about resources not created by Terraform (or by a disconnected Terraform setup).

**Reply by rgilmetdinov-i3** [reply]
Oh, it looks like I misunderstood how exactly to connect the provider. I used data to connect the provider to the desired organization, but it seems that’s not how it works.

Right now, my terraform plan shows:
```
  + zitadel_org                   = {
      + id             = (known after apply)
      + is_default     = null
      + name           = "test-org"
      + primary_domain = (known after apply)
      + state          = (known after apply)
    }
  + zitadel_project               = {
      + has_project_check        = null
      + id                       = (known after apply)
      + name                     = "my-test-project"
      + org_id                   = (known after apply)
      + private_labeling_setting = "PRIVATE_LABELING_SETTING_UNSPECIFIED"
      + project_role_assertion   = null
      + project_role_check       = null
      + state                    = (known after apply)
    }
```

Thank you @sm3142  so much for your help!

One more question: is it possible to connect the provider to an instance in the Terraform configuration without having to log in to the Zitadel console and create a service user?

**Reply by sm3142** [reply]
That looks all right for this simple example.

> One more question: is it possible to connect the provider to an instance in the Terraform configuration without having to log in to the Zitadel console and create a service user?

I wouldn't think so. You might be able to use the admin user created when launching the instance (can one create keys for human user accounts though?), but since this would require some manual bootstrapping also (creating a key), I would not recommend it. IMHO, it's better to create a dedicated service user with appropriate roles assigned and handle the key file with care (i.e. only make them available as secrets in the deployment pipelines, at least those for sensitive instances).

What I do when launching a new instance is to just use the default organization to host the (human) instance admins and a service user for Terraform IaC. Everything else then goes into dedicated, Terraform-managed organizations. I consider this to be the main selling point of ZITADEL Cloud over other AaaS offerings, i.e. the ability to create as many instances and organizations as you need, without breaking the bank or one's insanity. Anyway, this works pretty well for me and keeps the manual steps of creating a new instance to a minimum. But as always, YMMV.

**Reply by rgilmetdinov-i3** [reply]
I see, that makes sense. Thank you @sm3142 

That's weird but when I tried to run terraform apply via GitHub actions, I got almost the same issue:

```
Terraform will perform the following actions:
  # zitadel_org.test-org will be created
  + resource "zitadel_org" "test-org" {
      + id             = (known after apply)
      + name           = "test-org"
      + primary_domain = (known after apply)
      + state          = (known after apply)
    }
  # zitadel_project.test-project will be created
  + resource "zitadel_project" "test-project" {
      + id                       = (known after apply)
      + name                     = "my-test-project"
      + org_id                   = (known after apply)
      + private_labeling_setting = "PRIVATE_LABELING_SETTING_UNSPECIFIED"
      + state                    = (known after apply)
    }
Plan: 2 to add, 0 to change, 0 to destroy.
Changes to Outputs:
  + zitadel_org                   = {
      + id             = (known after apply)
      + is_default     = null
      + name           = "test-org"
      + primary_domain = (known after apply)
      + state          = (known after apply)
    }
  + zitadel_project               = {
      + has_project_check        = null
      + id                       = (known after apply)
      + name                     = "my-test-project"
      + org_id                   = (known after apply)
      + private_labeling_setting = "PRIVATE_LABELING_SETTING_UNSPECIFIED"
      + project_role_assertion   = null
      + project_role_check       = null
      + state                    = (known after apply)
    }
zitadel_org.test-org: Creating...
Error: rpc error: code = Unknown desc = unexpected HTTP status code received from server: 464 (); malformed header: missing HTTP content-type
  with zitadel_org.test-org,
  on zitadel.tf line 1, in resource "zitadel_org" "test-org":
   1: resource "zitadel_org" "test-org" {
```

I thought I might have forgotten to declare my service user token, but that’s not the case. Hmm, What may it be related with?

Oh, by the way, I forgot to mention that I ended up using the variable `jwt_profile_file = "./my_credentials.json"`, inserting the key file after I generated it.

**Reply by Rumbles** [reply]
@rgilmetdinov-i3 I seem to be seeing this same issue (Error: rpc error: code = Unknown desc = unexpected HTTP status code received from server: 464 (); malformed header: missing HTTP content-type) when I run zitadel in k8s behind an alb. I observe that if I do a port forward and point my provider at localhost instead the error is resolved.

Did you ever get to the bottom of it?

**Reply by Rumbles** [reply]
In my case setting this annotation on the alb fixed it:

    alb.ingress.kubernetes.io/backend-protocol-version: HTTP2

**Comment by fforootd** [comment]
CC @eliobischof maybe you have an idea here

**Reply by rgilmetdinov-i3** [reply]
anybody has some idea about this issue? I use the last zitadel provider version.

**Reply by sm3142** [reply]
As I mentioned before, I don't think I've ever seen this particular error before - and I use the ZITADEL Terraform provider pretty extensively. In my experience, the error messages when I muck up my Terraform spec (which happens with alarming frequency) are usually quite clear about the cause. It also looks like authentication for the provider has been set up correctly, otherwise it'd already fail at `terraform plan`. Really not sure what's going on here 🤔.

1. Are you running this against a ZITADEL cloud or a self-hosted instance?
1. Does `terraform apply` work when you run it locally, rather than in a Github workflow?
1. Can you post you Terraform spec in its entirety again?
